staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -397,10 +398,15 @@ func NewCacherFromConfig(config Config) (*Cacher, error) {
 		// so that future reuse does not get a spurious timeout.
 		<-cacher.timer.C
 	}
-	progressRequester := newConditionalProgressRequester(config.Storage.RequestWatchProgress, config.Clock)
+	// Add grpc context metadata to watch and progress notify requests done by cacher to:
+	// * Prevent starvation of watch opened by cacher, by moving it to separate Watch RPC than watch request that bypass cacher.
+	// * Ensure that progress notification requests are executed on the same Watch RPC as their watch, which is required for it to work.
+	contextMetadata := metadata.New(map[string]string{"source": "watch-cache"})
```

Comment:
Consider making this conditional on a gate so if we find a problem with it, we can disable it easily without respinning a build.

*If* we backport this (I'm on the fence there as well since this isn't a regression), we could default the gate off in backports as well to make the change opt-in

Reply:
Done, enabled by default. Any thoughts about release state?

Reply:
Marked as Beta

Reply:
I 100% agree that we need a gate here, but I would go directly to beta here, given we're effectively fixing a bug here.

Reply:
I think it is beta

Reply:
+1 to Beta